### PR TITLE
Add Governance to the constantProduct AMM

### LINKED ISSUE #3862: (After ConstantProduct AMM) add ParamGovernance to ConstantProduct AMM
## What is the Problem Being Solved?

The ConstantProduct AMM (#3771, #3791) was added in parallel to the governance framework (#3448, 3189). It needs to be updated to include governance functionality

## Description of the Design

The AMM's parameters should be visibly under the control of Governance.

## Security Considerations

Requires serious security review

## Test Plan

ensure governed params are visible.

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
All existing tests pass
Add tests for contract governance of Constant Product AMM
Also some internal documentation

closes: #3862

## Description

Puts the mutable parameters for the AMM under governance.

### Security Considerations

This is part of governance of the AMM, a key part of our economy, so the governance impacts need to be reviewed carefully. There's a  file `vpool-xyk-amm/governance.md` that describes the governance approach and considerations.

### Documentation Considerations

I will add contract documentation to the docs repository.

See https://github.com/Agoric/documentation/pull/566

### Testing Considerations

Tests updated, and new tests added.

### CODE REVIEW

File: packages/governance/src/contractGovernor.js [hunk 1]
```diff
@@ -155,6 +155,7 @@ const start = async (zcf, privateArgs) => {
   const creatorFacet = Far('governor creatorFacet', {
     voteOnParamChange,
     getCreatorFacet: () => limitedCreatorFacet,
+    getInternalCreatorFacet: () => E(governedCF).getInternalCreatorFacet(),
```
Comment (dckc) [inline packages/governance/src/contractGovernor.js:156]: I'd like to see a corresponding addition to the `GovernedContractFacetAccess` type declaration.

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  } = await E(zoe).startInstance(installs.electorate, {}, electorateTerms);
+
+  const governorTerms = {
+    timer,
+    electorateInstance,
+    governedContractInstallation: installs.autoswap,
+    governed: {
+      terms: ammTerms,
+      issuerKeywordRecord: { Central: centralR.issuer },
+      privateArgs: {},
```
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js]: This suggests `privateArgs` is required. Is it? It looks like it gets passed all the way through to zoe, so it could be left undefined.

It's missing from the informal type declaration in in `contractGovernor.js`:

> Those details for the governed contract are included in this contract's terms as a "governed" record. ...

not critical.

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+test('amm change param via Governance', async t => {
+  const centralR = makeIssuerKit('central');
+  const electorateTerms = { committeeName: 'EnBancPanel', committeeSize: 3 };
+  const timer = buildManualTimer(console.log);
+
+  const ammTerms = {
+    timer,
+    poolFeeBP: POOL_FEE_BP,
+    protocolFeeBP: PROTOCOL_FEE_BP,
+    main: ammInitialValues,
```
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js:176]: I wonder about renaming `main:` to `settings:` or `config:` or `state:` or some such to indicate terms that may vary over time.

not critical (and somewhat out of scope of this PR)

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+  const invitations = await E(committeeCreator).getVoterInvitations();
+  const { governorCreatorFacet } = governor;
+  const foo = await E(governorCreatorFacet).voteOnParamChange(
+    { parameterName: PROTOCOL_FEE_KEY },
+    20n,
+    installs.counter,
+    2n,
+  );
+  const { details } = foo;
```
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js]: I suggest not bothering with `foo`:

```suggestion
  const { details } = await E(governorCreatorFacet).voteOnParamChange(
    { parameterName: PROTOCOL_FEE_KEY },
    20n,
    installs.counter,
    2n,
  );
```

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
```
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js]: again:

```suggestion
  const { details } = await E(governorCreatorFacet).voteOnParamChange(
    { parameterName: PROTOCOL_FEE_KEY },
    20n,
    installs.counter,
    2n,
  );
```

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js [hunk 6]
```diff
@@ -30,25 +31,190 @@ import {
...
+    name: PROTOCOL_FEE_KEY,
+    value: 6n,
+    type: ParamType.NAT,
+  },
+]);
+
+test('amm with valid offers', async t => {
   // Set up central token
   const centralR = makeIssuerKit('central');
-  const centralTokens = value => AmountMath.make(value, centralR.brand);
```
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js:51]: wow... how did that ever work?

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js [hunk 7]
```diff
@@ -30,25 +31,190 @@ import {
...
 
-test('amm with valid offers', async t => {
-  const { moolaR, simoleanR, moola, simoleans } = setup();
-  const { zoeService } = makeZoeKit(fakeVatAdmin);
-  const feePurse = E(zoeService).makeFeePurse();
-  const zoe = E(zoeService).bindDefaultFeePurse(feePurse);
-  const invitationIssuer = await E(zoe).getInvitationIssuer();
-  const invitationBrand = await E(invitationIssuer).getBrand();
+  const {
+    creatorFacet: committeeCreator,
```
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js:131]: We don't actually use the committee in the tests in this file, right? We could have used the do-nothing electorate? I suppose it was convenient to just copy code that was known to work?

File: packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js [hunk 8]
```diff
@@ -50,7 +50,12 @@ test(`zoeHelper with zcf - swap`, async t => {
   assertPayoutAmount(t, simoleanIssuer, seat1PayoutB, simoleans(4));
   const seat2PayoutB = await bUserSeat.getPayout('B');
   assertPayoutAmount(t, simoleanIssuer, seat2PayoutB, simoleans(3));
-  assertPayoutAmount(t, moolaIssuer, await bUserSeat.getPayout('A'), moola(2));
+  await assertPayoutAmount(
+    t,
+    moolaIssuer,
+    await bUserSeat.getPayout('A'),
+    moola(2),
+  );
```
Comment (dckc) [inline packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js:58]: all the changes in this file seem to be whitespace / lint. odd. I wonder how that happened.

File: packages/governance/src/contractHelper.js [hunk 9]
```diff
@@ -38,6 +38,7 @@ const handleParamGovernance = (zcf, governedParamsTemplate) => {
       getGovernedParamsValues: () => {
         return { main: paramManager.getParams() };
       },
+      getParamValue: paramManager.getParam,
```
Comment (dckc) [inline packages/governance/src/contractHelper.js]: This API change seems like it merits an updated type declaration too.

darn; `HandleParamGovernance` doesn't include a `@returns`. Which means it defaults to `any`. That's another reason I don't like the distant `@callback` form. If we keep function type declarations with the functions, we get useful implicit return types.

File: packages/zoe/src/contracts/vpool-xyk-amm/governance.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+votes that would change the parameter values in a transparent way. When
+correctly set up, customers with access to this contract's publicFacet can
+verify the connectivity, and see which Electorate has the ability to vote on
+changes, and which votes are ongoing or have taken place. If not correctly set
+up, the validation checks will fail visibly.
+
+## Parameter Governance
+
+params.js declares PoolFee and ProtocolFee both Bigints, which
+can bu update by a governance vote. The contract gets the initial values for
```
Comment (dckc) [inline packages/zoe/src/contracts/vpool-xyk-amm/governance.md]:
```suggestion
can be update by a governance vote. The contract gets the initial values for
```

File: packages/governance/src/paramManager.js [hunk 11]
```diff
@@ -128,7 +128,9 @@ const buildParamManager = paramDescriptions => {
     });
     return harden(descriptions);
   };
-  const getParam = name => harden(makeDescription(name));
+  const getParam = name => {
+    return harden(makeDescription(name));
+  };
```
Comment (dckc) [inline packages/governance/src/paramManager.js]: why was this changed?

File: packages/zoe/src/contracts/vpool-xyk-amm/governance.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
+# Governance and the Multipool AMM
+
+This contract has two parameters (poolFee and protocolFee) that are
+managed under the governance framework. The contract calls
```
Comment (dckc) [inline packages/zoe/src/contracts/vpool-xyk-amm/governance.md]: Is it worth saying

a. why poolFee and protocolFee should be under governance rather than fixed at design-time or `startInstance` time?
b. risks associated with various decisions that governance might make? Which parties are at an advantage or disadvangage if `poolFee` is high vs. low? changing quickly vs. slowly?

File: packages/zoe/src/contracts/vpool-xyk-amm/governance.md [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+verify the connectivity, and see which Electorate has the ability to vote on
+changes, and which votes are ongoing or have taken place. If not correctly set
+up, the validation checks will fail visibly.
+
+## Parameter Governance
+
+params.js declares PoolFee and ProtocolFee both Bigints, which
+can bu update by a governance vote. The contract gets the initial values for
+those parameters from its terms, and thereafter can be seen to only use the
+values provided by the `getParamValue()` method returned by the paramManager.
```
Comment (dckc) [inline packages/zoe/src/contracts/vpool-xyk-amm/governance.md]: This "can be seen" bit is good.

File: packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js [hunk 14]
```diff
@@ -1,26 +1,30 @@
...
+import { handleParamGovernance } from '@agoric/governance/src';
+
 import { assertIssuerKeywords } from '../../contractSupport';
 import { makeAddPool } from './pool.js';
 import { makeMakeAddLiquidityInvitation } from './addLiquidity.js';
 import { makeMakeRemoveLiquidityInvitation } from './removeLiquidity.js';
 
 import '../../../exported.js';
 import { makeMakeCollectFeesInvitation } from '../newSwap/collectFees.js';
-import { makeMakeSwapInvitation } from './swap';
```
Comment (dckc) [inline packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js:15]: odd. I wonder how this survived The Great Extension Sweep.

File: packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js [hunk 15]
```diff
@@ -56,6 +60,20 @@ import { makeDoublePool } from './doublePool';
...
+ * with access to this contract's publicFacet can verify the connectivity, and
+ * see which Electorate has the ability to vote on changes, and which votes are
+ * ongoing or have taken place. If not correctly set up, the validation checks
+ * will fail visibly.
+ *
+ * The initial values of the parameters are provided as poolFeeBP and
+ * protocolFeeBP in terms. The poolFee is charged in RUN and each collateral, so
+ * it is provided as a bigint. The protocolFee is always charged in RUN, but the
+ * initial value is specified as a bigint for consistency.
+ *
```
Comment (dckc) [inline packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js:76]: Repeating this prose risks version skew. Add "see also" in both directions?

Or "As discussed in governance.md" here and "see also" in governance.md.

**Review by dckc — COMMENTED** [review]
Looks pretty good. I don't think I found anything critical, but I'd like you to respond to these comments before I approve.

File: packages/governance/src/contractGovernor.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractGovernor.js:156]: done

File: packages/governance/src/paramManager.js (hunk 11 above)
Comment (Chris-Hibbert) [reply packages/governance/src/paramManager.js]: probably to temporarily add logging. I'll revert it.

File: packages/governance/src/contractHelper.js (hunk 9 above)
Comment (Chris-Hibbert) [reply packages/governance/src/contractHelper.js]: Added more types

File: packages/zoe/src/contracts/vpool-xyk-amm/governance.md (hunk 10 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/vpool-xyk-amm/governance.md]: "can be updated".

File: packages/zoe/src/contracts/vpool-xyk-amm/governance.md (hunk 12 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/vpool-xyk-amm/governance.md]: I'm not sure what you're asking for. This particular doc is internal, so only people who review the code are likely to see it.  Those seem like discussion topics about policy for our economy rather than ones aimed at implementors, reviewers, or customers.

What trade-offs should be expanded on, and who should see them?

My answers to those questions would be 

a) we don't have any principled reasoning for particular values, and there's been discussion about changing the fees over time.

b) fees that are too high might reduce usage. Fees that are too low might not attract enough liquidity or repay investors enough.
Rapid changes would probably make it harder to attract customers and liquidity, but I wouldn't expect the availability of governance to lead to more changes.

These feel more like open-ended discussion topics than answers to me, and that's part of the reason these are subject to governance. What would you like to see captured here?

File: packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js (hunk 14 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js:15]: Yeah, I was surprised, too.

File: packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js (hunk 15 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js:76]: I'll drop the redundant prose here and just refer to the doc. Maybe it should all have been in the contract?  I dunno.

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js]: I expanded that description to include them.

`privateArgs` was added to Zoe relatively recently, and I realized that they were available and useful partway through building governance, so I was inconsistent in my treatment of them.

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js:176]: good suggestion.

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js]: a vestige of debugging.   I should mark these temporary changes better so I find them more reliably.  I know: I'll add `foo` to my presubmit check.  (or if there are too many instance of `foo` in real code, I'll choose a temporary name I can check for.)

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js (hunk 5 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-amm-governance.js]: done

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js (hunk 7 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js:131]: yup

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js (hunk 6 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js:51]: It was originally `AmountMath.make(value, brand)`. There was an explicit decision to change to `AmountMath.make(brand, value)`, with backward compatibility support. The outdated usages are supported but deprecated.  I've been updating gradually as I notice them.

File: packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js (hunk 8 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js:58]: Hmm.  Yeah, I don't see an intermediate state in my local history showing a temporary change, which would have triggered the reformatting. 

WebStorm wouldn't let me save the modified version (it cleaned it up automatically on save), so I had to use emacs to save the reverted version.

File: packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js (hunk 6 above)
Comment (dckc) [reply packages/zoe/test/unitTests/contracts/vpool-xyk-amm/test-xyk-amm-swap.js:51]: TIL.

File: packages/zoe/src/contracts/vpool-xyk-amm/governance.md (hunk 12 above)
Comment (dckc) [reply packages/zoe/src/contracts/vpool-xyk-amm/governance.md]: short version: sounds like no, it's not worth discussing such things here.

But that's what the name "governance.md" led me to expect: topics about policy for the economy.

If we're just documenting the fact that this software component is connected to another software component by 2 parameter widgets, then just sticking it inside the contract .js file would work for me.

File: packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js (hunk 15 above)
Comment (dckc) [reply packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js:76]: I have a mild preference for in the contract.

not critical.

**Review by dckc — APPROVED** [review]
Looks good.

I made some effort to verify that the changes are all covered by tests. Now that I mention it, that's another thing that computers could do for us, isn't it?
